### ISSUE TITLE
Bug when an inplace op is done on the output of the forward of a autograd.Function and this output is a view of an intermediary result of the forward.

### ISSUE THREAD

**Issue by sublee** [pytorch/pytorch#26546]
## 🐛 Bug

Gradients aren't backpropagated when all three conditions are met:

1. Use `torch.utils.checkpoint`.
2. Place `nn.InstanceNorm?d` right before any in-place module such as `nn.ReLU(inplace=True)`.
3. The checkpoint returns the result of the in-place module.

## To Reproduce

```python
import torch
from torch import nn
from torch.utils.checkpoint import checkpoint

x = torch.ones(3, 2, 1, requires_grad=True)

model = nn.Sequential(
    nn.InstanceNorm1d(2),
    nn.ReLU(inplace=True),
)
y = checkpoint(model, x)

y.norm().backward()

print(x.grad)
# Output: None
# Expected: tensor([[[0.],[0.]], ...])
```

`x` the input tensor doesn't receive gradients. `x.grad` is still `None`, which is not expected.

## Expected behavior

`x.grad` should be `torch.zeros(3, 2, 1)`. If we remove any of the conditions, the gradients are backpropagated correctly:

```python
# No in-place operation
model = nn.Sequential(
    nn.InstanceNorm1d(2),
    nn.ReLU(inplace=False),
)
y = checkpoint(model, x)

# No checkpoint
model = nn.Sequential(
    nn.InstanceNorm1d(2),
    nn.ReLU(inplace=True),
)
y = model(x)

# Not at the end of checkpoint
model = nn.Sequential(
    nn.InstanceNorm1d(2),
    nn.ReLU(inplace=True),
    nn.Conv1d(2, 2, 1),
)
y = checkpoint(model, x)

# Other norm
model = nn.Sequential(
    nn.BatchNorm1d(2),
    nn.ReLU(inplace=True),
)
y = checkpoint(model, x)
```

## Environment

```
Collecting environment information...
PyTorch version: 1.2.0
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 18.04.2 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: version 3.10.2

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration: 
GPU 0: Tesla P40
GPU 1: Tesla P40
GPU 2: Tesla P40
GPU 3: Tesla P40
GPU 4: Tesla P40
GPU 5: Tesla P40
GPU 6: Tesla P40
GPU 7: Tesla P40

Nvidia driver version: 410.104
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.2

Versions of relevant libraries:
[pip3] numpy==1.16.4
[pip3] torch==1.2.0
[pip3] torchgpipe==0.0.3
[pip3] torchvision==0.4.0
[conda] blas                      1.0                         mkl  
[conda] mkl                       2019.4                      243  
[conda] mkl-include               2019.4                      243  
[conda] mkl-service               2.0.2            py36h7b6447c_0  
[conda] mkl_fft                   1.0.12           py36ha843d7b_0  
[conda] mkl_random                1.0.2            py36hd81dba3_0  
[conda] torch                     1.2.0                    pypi_0    pypi
[conda] torchgpipe                0.0.3                     dev_0    <develop>
[conda] torchvision               0.4.0                    pypi_0    pypi
```

cc @ezyang @SsnL @albanD @zou3519 @gqchen

**Comment by colesbury** [comment]
Not sure who to route this to... @sublee are you planning to investigate this further?

cc @gchanan: do we have someone in FB who will maintain the checkpointing code?

**Comment by sublee** [comment]
@colesbury Yes, I do. I guess I can fix this issue by modifying `CheckpointFunction`.

**Comment by sublee** [comment]
I found the smallest pattern in [`at::instance_norm`](https://github.com/pytorch/pytorch/blob/v1.2.0/aten/src/ATen/native/Normalization.cpp#L462):

```
     ┌───────── checkpoint ─────────┐
x <--│-- f <-- view <-- in-place <--│-- loss
     └──────────────────────────────┘
```

It can be reproduced without InstanceNorm. We need only one `grad_fn`, one `view`, and one in-place operation in a checkpoint:

```python
def run_fn(input):
    out = input + 1             # add a grad_fn
    out = out.view(out.size())  # make a view
    out = out.relu_()           # apply in-place op
    return out

input_var = torch.rand(1, requires_grad=True)
out = checkpoint(run_fn, input_var)
out.backward()

assert input_var.grad is not None  # FAILS!
assert out.grad_fn.__class__.__name__ == 'CheckpointFunctionBackward'  # FAILS!
```

In the above example, `out.grad_fn` should be `CheckpointFunctionBackward`, but it is `AsStridedBackward` instead actually.

It seems to be simply fixed by detaching checkpoint outputs. But I'm not sure the checkpoint is the right place to fix this problem. Maybe tensor views or autograd history has to fix it instead. However, I'll send a PR on the checkpoint. Don't mind reject it If this approach is not correct.

**Comment by sublee** [comment]
I guess the combination of a tensor shaping function (`view()`, `t()`, `squeeze()`, `flatten()`, ...) and an in-place operation has a problem in autograd:

```python
x = torch.ones(1, requires_grad=True)

# EXPECTED:

# [1] No problem without a tensor shaping function.
y = x + 0
y.relu_()
>>> y
tensor([1.], grad_fn=<ReluBackward1>)

# [2] Under no_grad(), no problem without a tensor shaping function.
with torch.no_grad():
    y = x + 0
    y.relu_()
>>> y
tensor([1.])

# BUGS:

# [3] grad_fn should be ReluBackward1, but it is AsStridedBackward.
y = x.t()
y.relu_()
>>> y
tensor([1.], grad_fn=<AsStridedBackward>)

# [4] Under no_grad(), it still tracks grad functions.
with torch.no_grad():
    y = x.t()
    y.relu_()
>>> y
tensor([1.], grad_fn=<AsStridedBackward>)
```

**Comment by sublee** [comment]
This issue doesn't seem a problem at the checkpoint module. I think we should suppress `DifferentiableViewMeta` in `no_grad` mode:

```c++
inline Tensor as_view(const Tensor & base, Tensor tensor, bool is_differentiable = true) {
  auto base_var = Variable(base);
  if (base_var.is_view()) {
    base_var = base_var.base();
  }

  // SUGGESTION
  if (!GradMode::is_enabled()) {
    is_differentiable = false;
  }
  // END

  return make_variable_view(std::move(base_var), std::move(tensor), is_differentiable);
}
```

@colesbury Now I'll submit a PR with this fix. Let's continue to talk in the PR.

**Comment by gchanan** [comment]
I believe this is a known issue.  @albanD is there another issue tracking this?

**Comment by albanD** [comment]
Not really. #26546 tries to address this but has too many side effects at the moment as discussed in the PR.
Short answer is that there is a not so simple fix for this issue only. Potentially a simpler fix when we have a cleaner view of what we allow in the autograd.

**Comment by gchanan** [comment]
@albanD: I would suggest editing the title of the post.  You aren't going to remember that the issue is in-place ops on views in autograd by reading the title.

**Comment by albanD** [comment]
This is now raising the appropriate warning tellement the user what happens.
This will become an error in a future version when the deprecation cycle is done.
